File: python/signals_feed.py
"""Signals breadth-feed reader (Python).

The /v1/signals endpoint is the raw, strategy-agnostic breadth feed: every
$0.20-$20 small-cap that BullAlert's scanner validated this session (hundreds
of rows), as identity + catch-time + session only. This script paginates the
whole session pool with limit/offset and prints it grouped by session,
freshest-first.

Informational data feed, NOT financial advice and NOT a recommendation.

Run:  python signals_feed.py
"""
from __future__ import annotations

import sys
from collections import defaultdict

from bullalert_client import BullAlertClient, BullAlertError, load_dotenv

PAGE_SIZE = 50


def fetch_all_signals(client: BullAlertClient) -> list[dict]:
    all_rows: list[dict] = []
    offset = 0
    for _ in range(50):  # hard page cap (a session is hundreds of rows)
        res = client.signals(session="all", limit=PAGE_SIZE, offset=offset)
        rows = res.get("data", {}).get("signals", [])
        all_rows.extend(rows)
        if len(rows) < PAGE_SIZE:  # short page => last page
            break
        offset += PAGE_SIZE
    return all_rows


def main() -> int:
    load_dotenv()
    client = BullAlertClient()

    print("BullAlert -- validated candidate breadth feed (/v1/signals)")
    print("Strategy-agnostic data feed. Informational only, NOT financial advice.\n")

    rows = fetch_all_signals(client)
    if not rows:
        print("No validated signals right now (market may be closed). Try as_of= to replay a past session.")
        return 0

    buckets: dict[str, list[dict]] = defaultdict(list)
    for r in rows:
        buckets[r.get("session", "unknown")].append(r)

    labels = {"pre_market": "PRE-MARKET", "market": "REGULAR HOURS", "after_hours": "AFTER HOURS"}
    for session in ("pre_market", "market", "after_hours"):
        session_rows = buckets.get(session, [])
        if not session_rows:
            continue
        print(f"== {labels[session]} ({len(session_rows)}) ==")
        for r in session_rows:
            caught = (r.get("caught_at") or "unknown").replace("T", " ").replace(".000Z", "Z")
            print(f"  {r['ticker']:<8} caught {caught}")
        print("")

    print(f"Total validated candidates this session: {len(rows)}")
    print("This is breadth (the universe we screened), not a pick list. Do your own research.")
    return 0


if __name__ == "__main__":
    try:
        sys.exit(main())
    except BullAlertError as err:
        print(f"\n[{err.code or 'error'}] {err}", file=sys.stderr)
        sys.exit(1)

File: javascript/signals-feed.js
/**
 * Signals breadth-feed reader (JavaScript).
 *
 * The /v1/signals endpoint is the raw, strategy-agnostic breadth feed: every
 * $0.20-$20 small-cap that BullAlert's scanner validated this session
 * (hundreds of rows), as identity + catch-time + session only. This script
 * paginates the whole session pool with limit/offset and prints it grouped by
 * session, freshest-first.
 *
 *   Informational data feed, NOT financial advice and NOT a recommendation.
 *
 * Run:  node signals-feed.js
 */
import { loadEnv } from './lib/env.js';
import { BullAlertClient, BullAlertError } from './lib/client.js';

loadEnv();

const PAGE_SIZE = 50;

async function fetchAllSignals(client) {
  const all = [];
  let offset = 0;
  let metaSession = 'current';
  // Page until a page comes back shorter than the page size (the last page).
  for (let page = 0; page < 50; page += 1) {
    const res = await client.signals({ session: 'all', limit: PAGE_SIZE, offset });
    const rows = res.data?.signals ?? [];
    metaSession = res.meta?.session ?? metaSession;
    all.push(...rows);
    if (rows.length < PAGE_SIZE) break; // short page => done
    offset += PAGE_SIZE;
  }
  return { all, metaSession };
}

function group(rows) {
  const buckets = { pre_market: [], market: [], after_hours: [] };
  for (const r of rows) {
    (buckets[r.session] ?? (buckets[r.session] = [])).push(r);
  }
  return buckets;
}

async function main() {
  const client = new BullAlertClient();

  console.log('BullAlert — validated candidate breadth feed (/v1/signals)');
  console.log('Strategy-agnostic data feed. Informational only, NOT financial advice.\n');

  const { all } = await fetchAllSignals(client);

  if (all.length === 0) {
    console.log('No validated signals right now (market may be closed). Try ?as_of= to replay a past session.');
    return;
  }

  const buckets = group(all);
  const order = ['pre_market', 'market', 'after_hours'];
  const labels = { pre_market: 'PRE-MARKET', market: 'REGULAR HOURS', after_hours: 'AFTER HOURS' };

  for (const session of order) {
    const rows = buckets[session] ?? [];
    if (rows.length === 0) continue;
    console.log(`== ${labels[session]} (${rows.length}) ==`);
    for (const r of rows) {
      const when = r.caught_at ? new Date(r.caught_at).toISOString().replace('T', ' ').replace('.000Z', 'Z') : 'unknown';
      console.log(`  ${r.ticker.padEnd(8)} caught ${when}`);
    }
    console.log('');
  }

  console.log(`Total validated candidates this session: ${all.length}`);
  console.log('This is breadth (the universe we screened), not a pick list. Do your own research.');
}

main().catch((err) => {
  if (err instanceof BullAlertError) {
    console.error(`\n[${err.code ?? 'error'}] ${err.message}`);
    process.exit(1);
  }
  console.error('Unexpected error:', err);
  process.exit(1);
});
